# Lab 02: DataFrames

## #1
Download the planes dataset. Load it into Python and store it in a variable `df`.

In [1]:
import pandas as pd

In [2]:
# Load the data
# Path: up one directory, then in the "data" folder, then the file called planes.csv.
df = pd.read_csv('../data/planes.csv')

## #2
Print the first few rows of the DataFrame using the `.head()` method of `df`.

In [3]:
df.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

## #3
Take a look at a few attributes of `df`: `.shape`, `.index` and `.columns`. Do the contents of these attributes give you any information about the DataFrame?

In [4]:
df.shape

(3322, 9)

In [5]:
df.index

RangeIndex(start=0, stop=3322, step=1)

In [6]:
df.columns

Index(['tailnum', 'year', 'type', 'manufacturer', 'model', 'engines', 'seats',
       'speed', 'engine'],
      dtype='object')

We can see the data has 3322 rows and 9 columns.
Based on the column names, it looks like we have information about planes:
type, manufacturer, model, and more.

## #4
Check the data types of the columns of `df`.

In [7]:
df.dtypes

tailnum          object
year            float64
type             object
manufacturer     object
model            object
engines           int64
seats             int64
speed           float64
engine           object
dtype: object

## #5
Select just the `year` column using `df['year']`.
Store the result in a variable called `year`. Print it out.

In [8]:
year = df['year']

In [9]:
# Could also use `year.head()` for just first five rows.
year

0       2004.0
1       1998.0
2       1999.0
3       1999.0
4       2002.0
         ...  
3317    2002.0
3318    1992.0
3319    2002.0
3320    1992.0
3321    1992.0
Name: year, Length: 3322, dtype: float64

## #6
What is the type of `year`?

In [10]:
type(year)

pandas.core.series.Series

It's a Pandas `Series`.

## #7
Select *both* the `manufacturer` and `model` columns and store the result in a variable called `plane_types`.
Print it out.

In [11]:
plane_types = df[['manufacturer', 'model']]
plane_types

manufacturer      model
0                           EMBRAER  EMB-145XR
1                  AIRBUS INDUSTRIE   A320-214
2                  AIRBUS INDUSTRIE   A320-214
3                  AIRBUS INDUSTRIE   A320-214
4                           EMBRAER  EMB-145LR
...                             ...        ...
3317                         BOEING    717-200
3318  MCDONNELL DOUGLAS AIRCRAFT CO      MD-88
3319                         BOEING    717-200
3320  MCDONNELL DOUGLAS CORPORATION      MD-88
3321  MCDONNELL DOUGLAS CORPORATION      MD-88

[3322 rows x 2 columns]

## #8
What is the type of `plane_types`?
Why is this different from #6?

In [12]:
type(plane_types)

pandas.core.frame.DataFrame

`plane_types` is a Pandas `DataFrame`.
Because we used double-bracket notation (necessary to select multiple columns), we got back a `DataFrame` instead of a `Series`.

You could also figure this out because `Series` must be one-dimensional, but the result we got back is two-dimensional.

## #9
Slice the row at index 105.
What year was this plane made?

In [13]:
row_105 = df.loc[105]
row_105

tailnum                          N13161
year                               2004
type            Fixed wing multi engine
manufacturer                    EMBRAER
model                         EMB-145XR
engines                               2
seats                                55
speed                               NaN
engine                        Turbo-fan
Name: 105, dtype: object

In [14]:
row_105['year']

2004.0

The plane was made in 2004.

You could also do this in one line, if you're comfortable with the more compact syntax.

In [15]:
df.loc[105, 'year']

2004.0

## #10
Filter down the dataset to only planes made by Boeing.
How many rows are left?

In [16]:
boeing_planes = df[df['manufacturer'] == 'BOEING']
# Take a quick look to be sure it makes sense.
boeing_planes.head()

tailnum    year                     type manufacturer    model  engines  \
34  N11206  2000.0  Fixed wing multi engine       BOEING  737-824        2   
49  N1200K  1998.0  Fixed wing multi engine       BOEING  767-332        2   
50  N1201P  1998.0  Fixed wing multi engine       BOEING  767-332        2   
51  N12109  1994.0  Fixed wing multi engine       BOEING  757-224        2   
52  N12114  1995.0  Fixed wing multi engine       BOEING  757-224        2   

    seats  speed     engine  
34    149    NaN  Turbo-fan  
49    330    NaN  Turbo-fan  
50    330    NaN  Turbo-fan  
51    178    NaN  Turbo-jet  
52    178    NaN  Turbo-jet

In [17]:
boeing_planes.shape

(1630, 9)

So 1630 rows meet that criterion.

## #11
In the filtered data from #10, what do you notice about the indexes?
Call the `.reset_index()` method on this data and see what changes.

We can see that the index is no longer continuous:

In [18]:
boeing_planes.head()

tailnum    year                     type manufacturer    model  engines  \
34  N11206  2000.0  Fixed wing multi engine       BOEING  737-824        2   
49  N1200K  1998.0  Fixed wing multi engine       BOEING  767-332        2   
50  N1201P  1998.0  Fixed wing multi engine       BOEING  767-332        2   
51  N12109  1994.0  Fixed wing multi engine       BOEING  757-224        2   
52  N12114  1995.0  Fixed wing multi engine       BOEING  757-224        2   

    seats  speed     engine  
34    149    NaN  Turbo-fan  
49    330    NaN  Turbo-fan  
50    330    NaN  Turbo-fan  
51    178    NaN  Turbo-jet  
52    178    NaN  Turbo-jet

In [19]:
df = df.reset_index()
df

index tailnum    year                     type  \
0         0  N10156  2004.0  Fixed wing multi engine   
1         1  N102UW  1998.0  Fixed wing multi engine   
2         2  N103US  1999.0  Fixed wing multi engine   
3         3  N104UW  1999.0  Fixed wing multi engine   
4         4  N10575  2002.0  Fixed wing multi engine   
...     ...     ...     ...                      ...   
3317   3317  N997AT  2002.0  Fixed wing multi engine   
3318   3318  N997DL  1992.0  Fixed wing multi engine   
3319   3319  N998AT  2002.0  Fixed wing multi engine   
3320   3320  N998DL  1992.0  Fixed wing multi engine   
3321   3321  N999DN  1992.0  Fixed wing multi engine   

                       manufacturer      model  engines  seats  speed  \
0                           EMBRAER  EMB-145XR        2     55    NaN   
1                  AIRBUS INDUSTRIE   A320-214        2    182    NaN   
2                  AIRBUS INDUSTRIE   A320-214        2    182    NaN   
3                  AIRBUS INDUSTRIE   A320-214        2    182    NaN   
4                           EMBRAER  EMB-145LR        2     55    NaN   
...                             ...        ...      ...    ...    ...   
3317                         BOEING    717-200        2    100    NaN   
3318  MCDONNELL DOUGLAS AIRCRAFT CO      MD-88        2    142    NaN   
3319                         BOEING    717-200        2    100    NaN   
3320  MCDONNELL DOUGLAS CORPORATION      MD-88        2    142    NaN   
3321  MCDONNELL DOUGLAS CORPORATION      MD-88        2    142    NaN   

         engine  
0     Turbo-fan  
1     Turbo-fan  
2     Turbo-fan  
3     Turbo-fan  
4     Turbo-fan  
...         ...  
3317  Turbo-fan  
3318  Turbo-fan  
3319  Turbo-fan  
3320  Turbo-jet  
3321  Turbo-jet  

[3322 rows x 10 columns]

We can see that calling `.reset_index()` switches the index back to being simple increasing integers.

## #12
How many planes in the data have fewer than 40 seats?

In [20]:
small_planes = df[df['seats'] < 40]
small_planes.shape

(122, 10)

122 rows, so 122 planes

## #13
What is the minimum number of seats of any plane in the data?

In [21]:
df['seats'].min()

2

## #14
What is the minimum number of seats in any plane made by Boeing?

In [22]:
# We can use the boeing_planes dataset we created earlier.
boeing_planes['seats'].min()

100

... or we can start from scratch ...

In [23]:
# Note that we first filter rows (manufacturer == BOEING) and then select the column we want (seats)
df[df['manufacturer'] == 'BOEING']['seats'].min()

100

## #15
Create a new column in `df` called `seats_per_engine`.
It is defined as number of seats divided by number of engines.

In [24]:
df['seats_per_engine'] = df['seats'] / df['engines']
df.head()

index tailnum    year                     type      manufacturer  \
0      0  N10156  2004.0  Fixed wing multi engine           EMBRAER   
1      1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   
2      2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   
3      3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   
4      4  N10575  2002.0  Fixed wing multi engine           EMBRAER   

       model  engines  seats  speed     engine  seats_per_engine  
0  EMB-145XR        2     55    NaN  Turbo-fan              27.5  
1   A320-214        2    182    NaN  Turbo-fan              91.0  
2   A320-214        2    182    NaN  Turbo-fan              91.0  
3   A320-214        2    182    NaN  Turbo-fan              91.0  
4  EMB-145LR        2     55    NaN  Turbo-fan              27.5

## #16.
What is the average (mean) number of seats per engine in the data?

In [25]:
df['seats_per_engine'].mean()

77.06499598635361

## #17
Create a subset of the data, containing only planes manufactured by Cessna.
Select all the columns *except* manufacturer (since we know they're all Cessna).
Export the data as a CSV called "cessna_planes.csv".

In [26]:
# You can manually select columns:
cessna_planes = df[df['manufacturer'] == 'CESSNA'][['tailnum', 'year', 'type', 'model', 'engines', 'seats', 'speed', 'engine', 'seats_per_engine']]
cessna_planes.head()

tailnum    year                      type model  engines  seats  speed  \
424   N201AA  1959.0  Fixed wing single engine   150        1      2   90.0   
427   N202AA  1980.0   Fixed wing multi engine  421C        2      8   90.0   
893   N364AA  1973.0   Fixed wing multi engine  310Q        2      6  167.0   
1027  N378AA  1963.0  Fixed wing single engine  172E        1      4  105.0   
1476  N519AA  1979.0   Fixed wing multi engine   550        2      8    NaN   

             engine  seats_per_engine  
424   Reciprocating               2.0  
427   Reciprocating               4.0  
893   Reciprocating               3.0  
1027  Reciprocating               4.0  
1476      Turbo-fan               4.0

In [27]:
# Or you can "drop" just the manufacturer column explicitly:
cessna_planes = df[df['manufacturer'] == 'CESSNA'].drop('manufacturer', axis=1)
cessna_planes.head()

index tailnum    year                      type model  engines  seats  \
424     424  N201AA  1959.0  Fixed wing single engine   150        1      2   
427     427  N202AA  1980.0   Fixed wing multi engine  421C        2      8   
893     893  N364AA  1973.0   Fixed wing multi engine  310Q        2      6   
1027   1027  N378AA  1963.0  Fixed wing single engine  172E        1      4   
1476   1476  N519AA  1979.0   Fixed wing multi engine   550        2      8   

      speed         engine  seats_per_engine  
424    90.0  Reciprocating               2.0  
427    90.0  Reciprocating               4.0  
893   167.0  Reciprocating               3.0  
1027  105.0  Reciprocating               4.0  
1476    NaN      Turbo-fan               4.0

In [28]:
# Then, either way, save to a file.
cessna_planes.to_csv('../data/cessna_planes.csv')